src/tools/rbd_nbd/rbd-nbd.cc

```diff
@@ -825,86 +835,159 @@ static int parse_imgpath(const std::string &imgpath, Config *cfg)
   return 0;
 }
 
-static int do_list_mapped_devices()
+static int get_mapped_info(int pid, Config *cfg)
 {
-  char path[64];
-  int m = 0;
-  int fd[2];
+  int r;
+  std::string path = "/proc/" + stringify(pid) + "/cmdline";
+  std::ifstream ifs;
+  std::string cmdline;
+  std::vector<const char*> args;
 
-  if (socketpair(AF_UNIX, SOCK_STREAM, 0, fd) == -1) {
-    int r = -errno;
-    cerr << "rbd-nbd: socketpair failed: " << cpp_strerror(-r) << std::endl;
-    return r;
+  ifs.open(path.c_str(), std::ifstream::in);
+  assert (ifs.is_open());
+  ifs >> cmdline;
+
+  for (unsigned i = 0; i < cmdline.size(); i++) {
+    const char *arg = &cmdline[i];
+    if (i == 0) {
+      if (strcmp(basename(arg) , "rbd-nbd") != 0) {
+        return -EINVAL;
+      }
+    } else {
+      args.push_back(arg);
+    }
+ 
+    while (cmdline[i] != '\0') {
+      i++;
+    }
   }
+ 
+  std::ostringstream err_msg;
+  r = parse_args(args, &err_msg, cfg);
+  return r;
+}
 
-  while (true) {
-    snprintf(path, sizeof(path), "/dev/nbd%d", m);
-    int nbd = open_device(path);
-    if (nbd < 0)
-      break;
-    if (ioctl(nbd, NBD_SET_SOCK, fd[0]) != 0)
-      cout << path << std::endl;
-    else
-      ioctl(nbd, NBD_CLEAR_SOCK);
-    close(nbd);
-    m++;
+static int get_map_pid(const std::string& pid_path)
+{
+  int pid = 0;
+  std::ifstream ifs;
+  ifs.open(pid_path.c_str(), std::ifstream::in);
+  if (!ifs.is_open()) {
+    return 0;
   }
-
-  close(fd[0]);
-  close(fd[1]);
-
-  return 0;
+  ifs >> pid;
+  return pid;
 }
 
-static int rbd_nbd(int argc, const char *argv[])
+static int do_list_mapped_devices()
 {
   int r;
+  bool should_print = false;
+  int index = 0;
+  int pid = 0;
 
-  Config cfg;
+  std::string default_pool_name;
 
-  enum {
-    None,
-    Connect,
-    Disconnect,
-    List
-  } cmd = None;
+  TextTable tbl;
 
+  tbl.define_column("pid", TextTable::LEFT, TextTable::LEFT);
+  tbl.define_column("pool", TextTable::LEFT, TextTable::LEFT);
+  tbl.define_column("image", TextTable::LEFT, TextTable::LEFT);
+  tbl.define_column("snap", TextTable::LEFT, TextTable::LEFT);
+  tbl.define_column("device", TextTable::LEFT, TextTable::LEFT);
+ 
   vector<const char*> args;
+  auto cct = global_init(NULL, args, CEPH_ENTITY_TYPE_CLIENT,
+                         CODE_ENVIRONMENT_UTILITY, 0);
+ 
+  librados::Rados rados;
+  r = rados.init_with_context(g_ceph_context);
+  if (r < 0) {
+    cerr << "Failed to initialize RADOS: " << cpp_strerror(r) << std::endl;
+    return -1;
+  }
+  r = rados.connect();
+  if (r < 0) {
+    cerr << "Failed to connect to cluster: " << cpp_strerror(r) << std::endl;
+    return -1;
+  }
 
-  argv_to_vec(argc, argv, args);
-  md_config_t().parse_argv(args);
+  r = rados.conf_get("rbd_default_pool", default_pool_name);
```

Comment:
I think you don't need to initialise rados (moreover connect to the cluster) just to get "rbd_default_pool" config value. Also, global_init is not very good idea. Some time ago we changed the code to do global_init only for map command, to avoid error messages due to admin socket path collision.

You can use `md_config_t` directly, something like below:
```
md_config_t config();
config.parse_config_files(nullptr, nullptr, 0);
default_pool_name = config.rbd_default_pool;
```

There is still a possibility of incorrect result, if `rbd-nbd map` caller specified a different value for "rbd_default_pool" via CEPH_ARGS environment or command line args.  This can be improved a little if you will look for "rbd_default_pool" not here but in `parse_args()`. See the comment below.